### PR TITLE
feat: Add ChangeStreamRecordAdapter and ChangeStreamStateMachine

### PR DESCRIPTION
**Pull request by tengzhonger**
These will be used later for ChangeStreamMergingCallable.

### CODE REVIEW

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+          validate(
+              chunk.getMutation().hasSetCell(),
+              "AWAITING_CELL_VALUE: Current mod is not a SetCell.");
+          Mutation.SetCell setCell = chunk.getMutation().getSetCell();
+          numCellChunks++;
+          builder.cellValue(setCell.getValue());
+          // Case 1: Current SetCell is chunked.
+          if (chunk.hasChunkInfo()) {
+            validate(
+                chunk.getChunkInfo().getChunkedValueSize() != 0,
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:457]: nit: maybe validate (.. > 0) ?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+              validate(
+                  actualTotalSizeOfChunkedSetCell == chunk.getChunkInfo().getChunkedValueSize(),
+                  "Chunked value size in ChunkInfo doesn't match the actual total size. "
+                      + "ChunkInfo: "
+                      + chunk.getChunkInfo().getChunkedValueSize()
+                      + "; actual total size: "
+                      + actualTotalSizeOfChunkedSetCell);
+              return checkAndFinishMutationIfNeeded(dataChange, index + 1);
+            }
+            // If this is not the last chunk of a chunked SetCell, then this must be the last mod of
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:472]: Hmm not sure if I understand this comment, we could chat offline.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+  private State checkAndFinishMutationIfNeeded(
+      ReadChangeStreamResponse.DataChange dataChange, int index) {
+    validate(
+        0 <= index && index <= dataChange.getChunksCount(),
+        "checkAndFinishMutationIfNeeded: index out of bound.");
+    // Case 1): Handle the next mod.
+    if (index < dataChange.getChunksCount()) {
+      return AWAITING_NEW_MOD.handleMod(dataChange, index);
+    }
+    // We have finished handling all the mods in this DataChange.
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:539]: Should be 'in a logical mutation'

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+              validate(
+                  chunk.getChunkInfo().getChunkedValueOffset() == 0,
+                  "First chunk of a chunked cell must start with offset==0.");
+              actualTotalSizeOfChunkedSetCell = 0;
+            }
+            builder.startCell(
+                setCell.getFamilyName(),
+                setCell.getColumnQualifier(),
+                setCell.getTimestampMicros());
+            return AWAITING_CELL_VALUE.handleMod(dataChange, index);
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:392]: This is checked in the AWAITING_CELL_VALUE.handleMod.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ChangeStreamRecordAdapter.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+import com.google.bigtable.v2.ReadChangeStreamResponse;
+import com.google.cloud.bigtable.data.v2.models.Range.TimestampRange;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import javax.annotation.Nonnull;
+
+/**
+ * An extension point that allows end users to plug in a custom implementation of logical change
+ * stream records. This is useful in cases where the user would like to apply advanced client side
+ * filtering. This adapter acts like a factory for a SAX style change stream record builder.
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ChangeStreamRecordAdapter.java]: Discussed offline. Updated the comments to explain the potential use case.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java (hunk 1 above)
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java (hunk 3 above)
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java]: I realized that this is the right comment but is confusing. Updated the comments.  Does it make more sense now?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+              0 <= index && index <= dataChange.getChunksCount() - 1,
+              "AWAITING_CELL_VALUE: Index out of bound.");
+          ReadChangeStreamResponse.MutationChunk chunk = dataChange.getChunks(index);
+          validate(
+              chunk.getMutation().hasSetCell(),
+              "AWAITING_CELL_VALUE: Current mod is not a SetCell.");
+          Mutation.SetCell setCell = chunk.getMutation().getSetCell();
+          numCellChunks++;
+          builder.cellValue(setCell.getValue());
+          // Case 1: Current SetCell is chunked.
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:454]: For example: [ReadChangeStreamResponse1: {DC, DF, SC1}, ReadChangeStreamResponse2: {SC2, DF}]

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+              builder.finishCell();
+              validate(
+                  actualTotalSizeOfChunkedSetCell == chunk.getChunkInfo().getChunkedValueSize(),
+                  "Chunked value size in ChunkInfo doesn't match the actual total size. "
+                      + "ChunkInfo: "
+                      + chunk.getChunkInfo().getChunkedValueSize()
+                      + "; actual total size: "
+                      + actualTotalSizeOfChunkedSetCell);
+              return checkAndFinishMutationIfNeeded(dataChange, index + 1);
+            }
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:471]: maybe add else {
   validate( ... )
}
so the structure is cleaner?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+        }
+      };
+
+  /**
+   * A state that represents a completed change stream record. It prevents new change stream records
+   * from being read until the current one has been consumed.
+   *
+   * <dl>
+   *   <dt>Valid exit states:
+   *   <dd>{@link ChangeStreamStateMachine#AWAITING_NEW_STREAM_RECORD}.
```
Comment (mutianf) [inline google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:492]: how will this exit to AWAITING_NEW_STREAM_RECORD?

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+  boolean hasCompleteChangeStreamRecord() {
+    return currentState == AWAITING_STREAM_RECORD_CONSUME;
+  }
+  /**
+   * Checks if the state machine is in the middle of processing a change stream record.
+   *
+   * @return True If there is a change stream record in progress.
+   */
+  boolean isChangeStreamRecordInProgress() {
+    return currentState != AWAITING_NEW_STREAM_RECORD;
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:212]: Maybe it's not necessary? I see https://github.com/googleapis/java-bigtable/blob/8b120de58f0dfba3573ab696fb0e5375e917a00e/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/StateMachine.java#L193 only checks `currentState != AWAITING_NEW_ROW` but doesn't check `currentState != AWAITING_ROW_CONSUME` or `completeRow != null`.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 10]
```diff
@@ -0,0 +1,580 @@
...
+        currentState == AWAITING_STREAM_RECORD_CONSUME,
+        "Change stream record not ready to consume: " + currentState);
+    ChangeStreamRecordT changeStreamRecord = completeChangeStreamRecord;
+    reset();
+    return changeStreamRecord;
+  }
+
+  /** Checks if there is a complete change stream record to be consumed. */
+  boolean hasCompleteChangeStreamRecord() {
+    return currentState == AWAITING_STREAM_RECORD_CONSUME;
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:204]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java (hunk 8 above)
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:492]: The user is supposed to call `consumeChangeStreamRecord()` after the `hasCompleteChangeStreamRecord()` succeeds. `consumeChangeStreamRecord` will then call `reset()` and reset the currentState to `AWAITING_NEW_STREAM_RECORD`.

So this state itself won't transit to `AWAITING_NEW_STREAM_RECORD`. I updated the comments.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+
+/**
+ * A state machine to produce change stream records from a stream of {@link
+ * ReadChangeStreamResponse}. A change stream record can be a heartbeat, a close stream message or a
+ * logical mutation.
+ *
+ * <p>Note that there can be two types of chunking for a logical mutation:
+ *
+ * <ul>
+ *   <li>Non-SetCell chunking. For example, a logical mutation has two mods, where the first mod is
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:33]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ChangeStreamRecordAdapter.java [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+import com.google.bigtable.v2.ReadChangeStreamResponse;
+import com.google.cloud.bigtable.data.v2.models.Range.TimestampRange;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Timestamp;
+import javax.annotation.Nonnull;
+
+/**
+ * An extension point that allows end users to plug in a custom implementation of logical change
+ * stream records. This is useful in cases where the user would like to apply advanced client side
+ * filtering(for example, only keep DeleteFamily's in the mutations). This adapter acts like a
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ChangeStreamRecordAdapter.java:28]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 13]
```diff
@@ -0,0 +1,580 @@
...
+import com.google.bigtable.v2.ReadChangeStreamResponse;
+import com.google.bigtable.v2.ReadChangeStreamResponse.DataChange.Type;
+import com.google.cloud.bigtable.data.v2.models.ChangeStreamRecordAdapter.ChangeStreamRecordBuilder;
+import com.google.cloud.bigtable.data.v2.models.Range.TimestampRange;
+import com.google.common.base.Preconditions;
+
+/**
+ * A state machine to produce change stream records from a stream of {@link
+ * ReadChangeStreamResponse}. A change stream record can be a heartbeat, a close stream message or a
+ * logical mutation.
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:28]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 14]
```diff
@@ -0,0 +1,580 @@
...
+import com.google.bigtable.v2.Mutation;
+import com.google.bigtable.v2.ReadChangeStreamResponse;
+import com.google.bigtable.v2.ReadChangeStreamResponse.DataChange.Type;
+import com.google.cloud.bigtable.data.v2.models.ChangeStreamRecordAdapter.ChangeStreamRecordBuilder;
+import com.google.cloud.bigtable.data.v2.models.Range.TimestampRange;
+import com.google.common.base.Preconditions;
+
+/**
+ * A state machine to produce change stream records from a stream of {@link
+ * ReadChangeStreamResponse}. A change stream record can be a heartbeat, a close stream message or a
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:27]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java [hunk 15]
```diff
@@ -0,0 +1,580 @@
...
+     *
+     * @throws IllegalStateException If the subclass can't handle CloseStream events.
+     */
+    ChangeStreamStateMachine.State handleCloseStream(
+        ReadChangeStreamResponse.CloseStream closeStream) {
+      throw new IllegalStateException();
+    }
+
+    /**
+     * Accepts a new mod and transitions to the next state.
```
[the remark this answers is not in the document]
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:256]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java (hunk 6 above)
Comment (tengzhonger) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:454]: Done.

File: google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mutianf) [reply google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/changestream/ChangeStreamStateMachine.java:212]: I see, I guess depends on how you're using it in the merger: 
https://github.com/googleapis/java-bigtable/blob/8b120de58f0dfba3573ab696fb0e5375e917a00e/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/RowMerger.java#L97

**Review by mutianf — APPROVED** [review]

**Pull request merged**